internal/app/machined/pkg/controllers/block/internal/volumes/locate.go

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package volumes
+
+import (
+	"fmt"
+
+	"github.com/siderolabs/gen/value"
+	"github.com/siderolabs/gen/xerrors"
+	"github.com/siderolabs/go-blockdevice/v2/partitioning"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/cel/celenv"
+	"github.com/siderolabs/talos/pkg/machinery/resources/block"
+)
+
+// LocateAndProvision locates and provisions a volume.
+//
+// It verifies that the volume is either already created, and ready to be used,
+// or provisions it.
+//
+//nolint:gocyclo,cyclop
+func LocateAndProvision(logger *zap.Logger, volumeContext ManagerContext) error {
+	volumeType := volumeContext.Cfg.TypedSpec().Type
+
+	if volumeType == block.VolumeTypeTmpfs {
+		// tmpfs volumes are always ready
+		volumeContext.Status.Phase = block.VolumePhaseReady
+
+		return nil
+	}
+
+	// below for partition/disk volumes:
+	if value.IsZero(volumeContext.Cfg.TypedSpec().Locator) {
+		return fmt.Errorf("volume locator is not set")
+	}
+
+	// attempt to discover the volume
+	for _, dv := range volumeContext.DiscoveredVolumes {
+		matches, err := volumeContext.Cfg.TypedSpec().Locator.Match.EvalBool(celenv.VolumeLocator(), map[string]any{
+			"volume": dv,
+		})
+		if err != nil {
+			return fmt.Errorf("error evaluating volume locator: %w", err)
+		}
+
+		if matches {
+			volumeContext.Status.Phase = block.VolumePhaseLocated
+			volumeContext.Status.Location = dv.DevPath
+			volumeContext.Status.PartitionIndex = int(dv.PartitionIndex)
+			volumeContext.Status.ParentLocation = dv.ParentDevPath
+
+			volumeContext.Status.UUID = dv.Uuid
+			volumeContext.Status.PartitionUUID = dv.PartitionUuid
+			volumeContext.Status.Size = dv.Size
+
+			return nil
+		}
+	}
+
+	if !volumeContext.DevicesReady {
+		// volume wasn't located, but devices are not ready yet
```

Comment:
It took me some time to understand this comment. Maybe something like this would be more clear :)

```suggestion
		// volume wasn't located and devices are not ready yet, wait until they are ready
```